test/unit/test_qpu_client.py

```diff
@@ -49,240 +48,213 @@ def test_init__sets_processor_and_timeout(mock_engagement_manager: mock.MagicMoc
-        seconds_left=engagement_seconds_left,
+        seconds_left=seconds_left,
         credentials=engagement_credentials,
-        port=port,
+        port=1234,
     )
+    mock_engagement_manager.get_engagement.return_value = _engagement
 
-    @rpcq_server_with_auth.rpc_handler
-    def execute_qpu_request(request: rpcq.messages.QPURequest, user: str, priority: int):
-        time.sleep(engagement_seconds_left * 2)
-
-    with run_rpcq_server(rpcq_server_with_auth, port):
-        with raises(TimeoutError, match=f"Timeout on client tcp://localhost:{port}, method name execute_qpu_request"):
-            request = RunProgramRequest(
-                id="",
-                priority=0,
-                program="",
-                patch_values={},
-            )
-            qpu_client.run_program(request)
+    assert qpu_client._calculate_timeout(engagement=_engagement) == seconds_left
 
 
-def test_handles_contiguous_engagements(
+def test_run_program__retries_on_timeout(
     mock_engagement_manager: mock.MagicMock,
     engagement_credentials: EngagementCredentials,
-    rpcq_server_with_auth: rpcq.Server,
-    port: int,
+    mocker: MockerFixture,
 ):
-    """Test that a request crossing the boundary between two contiguous engagements can successfully complete."""
+    """Test that if a program times out, it will be retried.
 
+    A real-world example is a request crossing the boundary between two contiguous engagements.
+    """
+
+    # SETUP
+    processor_id = "some-processor"
+    job_id = "some-job"
     qpu_client = QPUClient(
-        quantum_processor_id="some-processor",
+        quantum_processor_id=processor_id,
         engagement_manager=mock_engagement_manager,
         request_timeout=1.0,
     )
-
-    first_engagement_seconds_left = qpu_client.timeout / 10
-    mock_engagement_manager.get_engagement.side_effect = [
-        engagement(
-            quantum_processor_id="some-processor",
-            seconds_left=first_engagement_seconds_left,
-            credentials=engagement_credentials,
-            port=port,
-        ),
-        engagement(
-            quantum_processor_id="some-processor",
-            seconds_left=9999,
+    mock_engagement_manager.get_engagement.return_value = engagement(
+            quantum_processor_id=processor_id,
+            seconds_left=0,
             credentials=engagement_credentials,
-            port=port,
-        ),
+            port=1234,
+        )
+    client = patch_rpc_client(mocker=mocker, return_value=None)
+    client.call.side_effect = [
+        TimeoutError,  # First request must look like it timed out so we can verify retry
+        job_id,
     ]
+    request_kwargs = {"id": "TestingContiguous", "patch_values": {}, "program": ""}
+    request = RunProgramRequest(  # Thing we give to QPUClient
+        priority=0,
+        **request_kwargs,
+    )
+    qpu_request = QPURequest(**request_kwargs)  # Thing QPUClient gives to rpcq.Client
```

Comment:
Might consider moving this to just before the `client.call.assert_has_calls` assertion